### PR TITLE
Fix dns memoryleak by upgrading outdated miekg/dns

### PR DESCRIPTION
**Pull request by 42wim**
Found a memoryleak in the TCP DNS serving. We were seeing gradual memory build-up on our consul servers, up to 8GB until we restarted the process. (running v1.6.1)

After some debugging the problem seemed to be in `serveTCP`

```
(pprof) text
Showing nodes accounting for 353.40MB, 94.72% of 373.12MB total
Dropped 107 nodes (cum <= 1.87MB)
Showing top 10 nodes out of 71
      flat  flat%   sum%        cum   cum%
   81.51MB 21.85% 21.85%    81.51MB 21.85%  net.newFD
      73MB 19.57% 41.41%       73MB 19.57%  syscall.anyToSockaddr
   57.25MB 15.34% 56.76%    57.25MB 15.34%  github.com/hashicorp/go-msgpack/codec.(*msgpackDecDriver).decodeBytes
   51.50MB 13.80% 70.56%    51.50MB 13.80%  net.sockaddrToTCP
   26.50MB  7.10% 77.66%       38MB 10.19%  github.com/hashicorp/go-immutable-radix.(*Txn).insert
   19.12MB  5.13% 82.79%   228.14MB 61.14%  github.com/miekg/dns.(*Server).serveTCP
   13.50MB  3.62% 86.41%    13.50MB  3.62%  github.com/hashicorp/go-memdb.(*StringFieldIndex).FromObject
   11.50MB  3.08% 89.49%    11.50MB  3.08%  github.com/hashicorp/go-msgpack/codec.(*msgpackDecDriver).decodeString
      10MB  2.68% 92.17%       10MB  2.68%  github.com/hashicorp/go-immutable-radix.(*Txn).writeNode
    9.50MB  2.55% 94.72%   126.76MB 33.97%  github.com/hashicorp/consul/agent/consul/fsm.restoreKV
```
This issue has been mentioned in https://github.com/miekg/dns/issues/807 and closed in https://github.com/miekg/dns/pull/808

After deploying consul with updated dependencies the memory usage on our quality/production clusters stays stable.

### CODE REVIEW

**Comment by 42wim** [comment]
Also added tests and fixed an overflow issue when trimming UDP responses

File: agent/dns.go [hunk 1]
```diff
@@ -1077,7 +1077,7 @@ func trimUDPResponse(req, resp *dns.Msg, udpAnswerLimit int) (trimmed bool) {
 	// uncompresses them.
 	// Even when size is too big for one single record, try to send it anyway
 	// (useful for 512 bytes messages)
-	for len(resp.Answer) > 1 && resp.Len() > maxSize {
+	for len(resp.Answer) > 1 && resp.Len() > maxSize-7 {
```
Comment (pierresouchay) [inline agent/dns.go]: why 7?

Comment (42wim) [reply agent/dns.go]: https://github.com/miekg/dns/blob/9b7437f11d75979ac166ebd0780edd96e62d5145/dns.go#L64-L70
```
type RR_Header struct {
	Name     string `dns:"cdomain-name"`
	Rrtype   uint16
	Class    uint16
	Ttl      uint32
	Rdlength uint16 // Length of data after header.
}
```
The overflow fails on the TTL read, which is 4 bytes, the unpacking code checks if we have 4 bytes extra to read otherwise overflows and after that we still need 2 bytes for Rrdlength.
So the total is 6 bytes

**Comment by 42wim** [comment]
Anyone from hashicorp that can take a look and/or comment?
cc: @mkeeler

**Comment by i0rek** [comment]
This is the diff between the two versions: https://github.com/miekg/dns/compare/v1.0.14...v1.1.22.
